### PR TITLE
Issue21 Enhance visualization to show call stack 

### LINKED ISSUE #21: Enhance code window to show call stack
The current code window in VisualzationApp can only show one chunk of code at a time.   To show multiple call levels requires managing an explicit call stack.  That requires changing how the `self.cleanup` mechanism operates as well as the code window.
I put a design document in https://github.com/JMCanning78/datastructures-visualization/blob/JMCanning78-call-stack-design/PythonVisualizations/design/CallStackVisualization.md
This issue will track development on that design.

### PR DESCRIPTION
**Pull request by JMCanning78**
This is a significant change to the API of the VisualizationApp.  It allows for visualization to be controlled from separate calls on the call stack.  For example, the `delete()` method of Array.py calls the `find()` method to find the item to delete.  Both `delete()` and `find()` have local variables that should be shown in the visualizations, along with highlighting the code.  The two method calls can now be shown as two chunks of code in the code window and their local variables can be managed in separate environments.  (Array.py does not yet show code in this branch of the repo, and SimpleSorting.py doesn't create two levels in the call stack, but does show its code).

I've tested the branch pretty thoroughly.  It changes the way the visualization modules create their local environment, how that local environment is cleaned up, and interactions with the play/pause/stop controls.  I'd like a couple of other teams to review this PR by end-user testing it.  Future work on visualization will need to use the revised API to VisualizationApp.  Once tested, this will resolve #21, and everyone will need to migrate other modules toward this structure.

### CODE REVIEW

File: PythonVisualizations/SimpleSorting.py [hunk 1]
```diff
@@ -167,19 +166,18 @@ def insert(self, item):
     
     def insert(self, val):
         self.startAnimations()
-        self.cleanUp()
-        self.showCode(self.insertCode.strip())
-        self.createCodeTags(self.insertCodeSnippets)
+        callEnviron = self.createCallEnvironment(
+            self.insertCode.strip(), self.insertCodeSnippets)
```
Comment (lilypolonetsky) [inline PythonVisualizations/SimpleSorting.py:170]: doesn't `createCallEnvironment`  create a `CodeHighlightBlock` object which calls `code.strip()`? Is it necessary to call `strip()`  on `insertCode`?

Comment (JMCanning78) [reply PythonVisualizations/SimpleSorting.py:170]: Yes that `.strip()` in the arguments to the call is redundant.  It won't hurt anything to leave it either.  Modifying the branch and updating the PR, however, takes a bunch of time.

**Comment by JMCanning78** [comment]
@lilypolonetsky I have merged your changes for the Traverse operation, Natania & Esther's changes for the global variables into this branch.  Please re-review to see if it's ready for committing to master.

**Comment by JMCanning78** [comment]
Adding the global variables and converting `traverse()` to use the call stack API were necessary to resolve the conflicts from the various changes.  That's "normal" for this kind of merge.

When you review a PR, it's best to choose "Approve" or "Request Changes" instead of "Comment".  I'm going to go ahead with merging this PR even though it doesn't have an approval.

**Pull request merged**